include/kernels/madwf_tensor.cuh

```diff
@@ -0,0 +1,205 @@
+#pragma once
+
+#include <color_spinor_field.h>
+#include <color_spinor_field_order.h>
+#include <shared_memory_cache_helper.cuh>
+#include <kernels/madwf_transfer.cuh>
+#include <block_reduce_helper.h>
+
+namespace quda
+{
+
+  namespace madwf_ml
+  {
+    template <class T> __device__ __host__ inline void block_reduce_x(T &f)
+    {
+      int lane_id_x = target::thread_idx().x % device::warp_size();
+      int warp_id_x = target::thread_idx().x / device::warp_size();
+      int warp_ct_x = target::block_dim().x / device::warp_size();
+
+      // pad = false, dynamic = false, basically to get a __shared__ T smem[32];
+      SharedMemoryCache<T, 32, 1, false, false> cache;
+      T *smem = cache.data();
+
+      WarpReduce<T, device::warp_size()> warp_reduce;
+      f = warp_reduce.Sum(f);
+      // Now lane 0 of each warp holds the reduced value
+
+      if (warp_ct_x > 1) {
+        if (lane_id_x == 0) { smem[target::thread_idx().y * warp_ct_x + warp_id_x] = f; }
+        cache.sync();
+        if (warp_id_x == 0) {
+          f = (lane_id_x < warp_ct_x) ? smem[target::thread_idx().y * warp_ct_x + lane_id_x] : 0;
+          f = warp_reduce.Sum(f);
+        }
+      }
+      // Now the first thread in the x direction holds the reduction result.
+    }
+
+    /**
+      @brief Form a 4-by-4 tensor from v and w (color d.o.f. is contracted)
+      @param[in] mp The buffer to accumulate on
+      @param[in] v, w input vectors
+     */
+    template <class real>
+    __device__ __host__ inline void vector_tensor_matrix(SpinMatrix<real> *mp, int m_index, const WilsonVector<real> &v,
+                                                         const WilsonVector<real> &w)
+    {
+      complex<real> *p = reinterpret_cast<complex<real> *>(mp);
```

Comment:
Do we need the `reinterpret_cast`?  Can we not just use the matrix accessors directly?

Reply:
We do need the `reinterpret_cast` here since we have multiple matrices and we are using a strided access pattern.

Reply:
I was about to type this
> The concern with the `reinterpret_cast` is that it can lead to undefined behaviour, e.g., if the compiler inserts padding in the matrix.  A good test for this is often to see if the code runs ok in debug mode (`-G`).  In almost all circumstances we should `reinterpret_cast` on stack variables at all times.
But now I see when looking closer at the code that this is a global buffer and will be allocated on the heap.  So this will be safe to do, and I can see the point of this 😄